### PR TITLE
Openvino hybrid awareness

### PR DESCRIPTION
**Pull request by myshevts**
this is **OV hybrid awareness** that most of you already reviewed and approved before (as https://github.com/openvinotoolkit/openvino/pull/3199).
Now it is rebased (and peeled from oneTBB) to the branch (by @ivankochin) that backports the [**hybrid TBB APIs**  ](https://github.com/openvinotoolkit/openvino/pull/4602) to the OV.
Since Ivan's PR will be merged, I did a final align to the master and now inviting everyone for a review.

### CODE REVIEW

File: inference-engine/src/inference_engine/os/lin/lin_system_conf.cpp [hunk 1]
```diff
@@ -81,7 +82,16 @@ int getNumberOfCPUCores() {
...
         }
     }
-    return CPU_COUNT(&currentCoreSet);
+    int phys_cores = CPU_COUNT(&currentCoreSet);
+    #if (IE_THREAD == IE_THREAD_TBB || IE_THREAD == IE_THREAD_TBB_AUTO)
+    auto core_types = custom::info::core_types();
+    if (bigCoresOnly && core_types.size() > 1) /*Hybrid CPU*/ {
+        const auto little_cores = core_types.front();
+        // assuming the Little cores feature no hyper-threading
+        phys_cores -= custom::info::default_concurrency(custom::task_arena::constraints{}.set_core_type(little_cores));
```
Comment (ivankochin) [inline inference-engine/src/inference_engine/os/lin/lin_system_conf.cpp:91]: The following call returns the default concurrency for big cores only without hyperthreading (with respect to the process affinity)

```suggestion
        phys_cores = custom::info::default_concurrency(custom::task_arena::constraints{}
            .set_core_type(core_types.back())
            .set_max_threads_per_core(1)
        );
```

File: inference-engine/src/inference_engine/threading/ie_istreams_executor.cpp [hunk 2]
```diff
@@ -135,12 +134,52 @@ Parameter IStreamsExecutor::Config::GetConfig(const std::string& key) {
     return {};
 }
 
-IStreamsExecutor::Config IStreamsExecutor::Config::MakeDefaultMultiThreaded(const IStreamsExecutor::Config& initial) {
+IStreamsExecutor::Config IStreamsExecutor::Config::MakeDefaultMultiThreaded(const IStreamsExecutor::Config& initial, const bool fp_intesive) {
```
Comment (myshevts) [inline inference-engine/src/inference_engine/threading/ie_istreams_executor.cpp:137]: logic to select core-types

File: inference-engine/src/mkldnn_plugin/config.cpp [hunk 3]
```diff
@@ -13,24 +13,32 @@
...
-#else
-    streamExecutorConfig._threadBindingType = InferenceEngine::IStreamsExecutor::NONE;
-#endif
-#else
+    // this is default mode
     streamExecutorConfig._threadBindingType = InferenceEngine::IStreamsExecutor::CORES;
-#endif
+
+    // for the TBB code-path, additional configuration depending on the OS and CPU types
+    #if (IE_THREAD == IE_THREAD_TBB || IE_THREAD == IE_THREAD_TBB_AUTO)
```
Comment (myshevts) [inline inference-engine/src/mkldnn_plugin/config.cpp:27]: detecting hybrid CU and making the right pinning deafult

File: inference-engine/samples/benchmark_app/benchmark_app.hpp [hunk 4]
```diff
@@ -64,10 +64,11 @@ static const char custom_cldnn_message[] = "Required for GPU custom kernels. Abs
 static const char batch_size_message[] = "Optional. Batch size value. If not specified, the batch size value is determined from Intermediate Representation.";
 
 // @brief message for CPU threads pinning option
-static const char infer_threads_pinning_message[] = "Optional. Enable threads->cores (\"YES\", default), threads->(NUMA)nodes (\"NUMA\") " \
-                                                    "or completely disable (\"NO\") " \
-                                                    "CPU threads pinning for CPU-involved inference.";
-
+static const char infer_threads_pinning_message[] = "Optional. Explicit inference threads binding options (leave empty to let the OpenVINO to make a choice):\n" \
```
Comment (myshevts) [inline inference-engine/samples/benchmark_app/benchmark_app.hpp:67]: adding new pinning type

File: inference-engine/samples/benchmark_app/main.cpp [hunk 5]
```diff
@@ -267,9 +267,6 @@ int main(int argc, char* argv[]) {
                     if ((device_name.find("MULTI") != std::string::npos) && (device_name.find("GPU") != std::string::npos)) {
                         slog::warn << "Turn off threads pinning for " << device << " device since multi-scenario with GPU device is used." << slog::endl;
                         device_config[CONFIG_KEY(CPU_BIND_THREAD)] = CONFIG_VALUE(NO);
-                    } else {
-                        // set to default value
-                        device_config[CONFIG_KEY(CPU_BIND_THREAD)] = FLAGS_pin;
```
Comment (myshevts) [inline inference-engine/samples/benchmark_app/main.cpp:272]: with introduction of the hybrid processors, our default pinning is not CORES anymore (it is HYBRID_AWARE on the hybrid processors)

File: inference-engine/src/inference_engine/ie_system_conf.cpp [hunk 6]
```diff
@@ -90,16 +90,25 @@ bool checkOpenMpEnvVars(bool includeOMPNumThreads) {
 #if defined(__APPLE__)
 // for Linux and Windows the getNumberOfCPUCores (that accounts only for physical cores) implementation is OS-specific
 // (see cpp files in corresponding folders), for __APPLE__ it is default :
-int getNumberOfCPUCores() { return parallel_get_max_threads();}
+int getNumberOfCPUCores(bool) { return parallel_get_max_threads();}
 #if !((IE_THREAD == IE_THREAD_TBB) || (IE_THREAD == IE_THREAD_TBB_AUTO))
-std::vector<int> getAvailableNUMANodes() { return {0}; }
+std::vector<int> getAvailableNUMANodes() { return {-1}; }
```
Comment (myshevts) [inline inference-engine/src/inference_engine/ie_system_conf.cpp:95]: aligned to the default fallback values that the TBB produces (e.g. when hwloc fails)

File: inference-engine/src/plugin_api/ie_system_conf.h [hunk 7]
```diff
@@ -36,13 +36,20 @@ INFERENCE_ENGINE_API_CPP(bool) checkOpenMpEnvVars(bool includeOMPNumThreads = tr
...
+INFERENCE_ENGINE_API_CPP(std::vector<int>) getAvailableCoresTypes();
+
 /**
  * @brief      Returns number of CPU physical cores on Linux/Windows (which is considered to be more performance friendly for servers)
  *             (on other OSes it simply relies on the original parallel API of choice, which usually uses the logical cores )
  * @ingroup    ie_dev_api_system_conf
  * @return     Number of physical CPU cores.
  */
-INFERENCE_ENGINE_API_CPP(int) getNumberOfCPUCores();
+INFERENCE_ENGINE_API_CPP(int) getNumberOfCPUCores(bool bigCoresOnly = false);
```
Comment (ilya-lavrenov) [inline inference-engine/src/plugin_api/ie_system_conf.h:52]: add a description for new variable.

**Review by ilya-lavrenov — APPROVED** [review]

File: inference-engine/src/inference_engine/os/win/win_system_conf.cpp [hunk 8]
```diff
@@ -32,12 +32,28 @@ int getNumberOfCPUCores() {
         offset += reinterpret_cast<PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX>(ptr.get() + offset)->Size;
         phys_cores++;
     } while (offset < sz);
+
+    #if (IE_THREAD == IE_THREAD_TBB || IE_THREAD == IE_THREAD_TBB_AUTO)
+    auto core_types = custom::info::core_types();
+    // TODO: REMOVE THE DEBUG PRINTF
+    for (auto tp : core_types) {
+        printf("type: %d %d concurency\n", tp, custom::info::default_concurrency(custom::task_arena::constraints{}.set_core_type(tp)));
+    }
```
Comment (ilya-lavrenov) [inline inference-engine/src/inference_engine/os/win/win_system_conf.cpp:41]: remove all places like `TODO: REMOVE THE DEBUG PRINTF`?

File: inference-engine/src/inference_engine/threading/ie_cpu_streams_executor.cpp [hunk 9]
```diff
@@ -67,39 +67,66 @@ struct CPUStreamsExecutor::Impl {
...
-                });
+                if (Config::PreferredCoreType::ROUND_ROBIN != _impl->_config._threadPreferredCoreType) {
+                   if (Config::PreferredCoreType::NONE == _impl->_config._threadPreferredCoreType) {
+                       // TODO: REMOVE THE DEBUG PRINTF
+                       printf("%s, NO BINDING, StreamId: %d (%d threads) \n",
+                           _impl->_config._name.c_str(), _streamId, _impl->_config._threadsPerStream);
+                   } else {
+                       const auto selected_core_type = Config::PreferredCoreType::BIG == _impl->_config._threadPreferredCoreType
+                           ? custom::info::core_types().back() // runing on Big cores only
+                           : custom::info::core_types().front(); // runing on Little cores only
```
Comment (ilya-lavrenov) [inline inference-engine/src/inference_engine/threading/ie_cpu_streams_executor.cpp:84]: running

**Review by apankratovantonp — APPROVED** [review]

**Review by ilya-lavrenov — APPROVED** [review]

File: inference-engine/samples/benchmark_app/README.md [hunk 10]
```diff
@@ -104,14 +104,16 @@ Options:
                                 estimations the number of streams should be set to 1.
     -nthreads "<integer>"       Optional. Number of threads to use for inference on the CPU (including HETERO and MULTI cases).
     -enforcebf16="<true/false>" Optional. By default floating point operations execution in bfloat16 precision are enforced if supported by platform.
-                                        'true'  - enable  bfloat16 regardless of platform support
-                                        'false' - disable bfloat16 regardless of platform support.
-    -pin "YES"/"NO"/"NUMA"      Optional. Enable threads->cores ("YES", default), threads->(NUMA)nodes ("NUMA") or completely disable ("NO") CPU threads pinning for CPU-involved inference.
+    -pin "YES"/"HYBRID_AWARE"NUMA"/"NO"
```
[the remark this answers is not in the document]
Comment (myshevts) [reply inference-engine/samples/benchmark_app/README.md:107]: HYBRID_AWARE affects only load-time cores selection, so there will be no additional volatility etc e.g. for the latency. but overall hybrid machiens are quite volatile (as entire idea is to aggressively save on power)

File: inference-engine/src/plugin_api/threading/ie_istreams_executor.hpp [hunk 11]
```diff
@@ -85,6 +87,12 @@ class INFERENCE_ENGINE_API_CLASS(IStreamsExecutor) : public ITaskExecutor {
         int                _threadBindingStep       = 1;  //!< In case of @ref CORES binding offset type thread binded to cores with defined step
         int                _threadBindingOffset     = 0;  //!< In case of @ref CORES binding offset type thread binded to cores starting from offset
         int                _threads                 = 0;  //!< Number of threads distributed between streams. Reserved. Should not be used.
+        enum PreferredCoreType {
+            NONE,
+            LITTLE,
+            BIG,
+            ROUND_ROBIN // for multiple streams
```
[the remark this answers is not in the document]
Comment (myshevts) [reply inference-engine/src/plugin_api/threading/ie_istreams_executor.hpp:94]: this is is indeed a hint for the cores types selection scheme ( populate big cores, then little, then wrap around if more streams are required). although non-static (depedns on number of the requested streams) , the selection is fully load-time though.
I will update the comment

File: inference-engine/src/plugin_api/ie_system_conf.h [hunk 12]
```diff
@@ -36,13 +36,21 @@ INFERENCE_ENGINE_API_CPP(bool) checkOpenMpEnvVars(bool includeOMPNumThreads = tr
...
+ * @ingroup    ie_dev_api_system_conf
+ * @return     Vector of core types
+ */
+INFERENCE_ENGINE_API_CPP(std::vector<int>) getAvailableCoresTypes();
+
 /**
  * @brief      Returns number of CPU physical cores on Linux/Windows (which is considered to be more performance friendly for servers)
  *             (on other OSes it simply relies on the original parallel API of choice, which usually uses the logical cores )
  * @ingroup    ie_dev_api_system_conf
+ * @param[in]  bigCoresOnly Additionally limits the number of reported cores to the 'Big' cores only
```
[the remark this answers is not in the document]
Comment (myshevts) [reply inference-engine/src/plugin_api/ie_system_conf.h:50]: all = call the func with false
big = call the func with true
little = all - big

i will add a comment

**Pull request merged**